### PR TITLE
finishing math mode

### PR DESCRIPTION
**Pull request by Mikhail-M**

### CODE REVIEW

**Comment by dbarashev** [comment]
Please follow the convention of naming branches. Normally branch should be named `tkt_[TICKET_NUMBER]_[MNEMONICS]`

where `TICKET_NUMBER` is a number of a ticket which the branch is related to

File: lib/ace/ext/papeeria/latex_parsing_context.coffee [hunk 1]
```diff
@@ -9,24 +9,26 @@ define((require, exports, module) ->
      *
      * Returns context at row.
     ###
-    getContext = (session, row) ->
+    # Specific for token"s system of type in ace
+  # We saw such a realization in html_completions.js
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/latex_parsing_context.coffee]: nit: fix indent

File: lib/ace/ext/papeeria/latex_parsing_context.coffee [hunk 2]
```diff
@@ -9,24 +9,26 @@ define((require, exports, module) ->
...
+        return token.type.split(".").indexOf(type) > -1
+
+    getContext = (session, row, column) ->
+        token = session.getTokenAt(row, column)
+        if token? and isType(token, "math")
+            return EQUATION_STATE
+        else
+            getContext_(session, row)
+
+    getContext_ = (session, row) ->
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/latex_parsing_context.coffee]: `getContextFromPos` and `getContextFromRow` must be better names than `getContext` and `getContext_`

File: lib/ace/ext/papeeria/snippets/equation_snippets.coffee [hunk 3]
```diff
@@ -0,0 +1,130 @@
+define((require, exports, module) ->
+  equationEnvironments = [
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/snippets/equation_snippets.coffee]: These array look like constants, so please rename them in constants style: `EQUATION_ENVIRONMENTS` etc

File: lib/ace/ext/papeeria/snippets/equation_snippets.coffee [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    "\\pi"
+    "\\sigma"
+    "\\upsilon"
+    "\\phi"
+    "\\chi"
+    "\\psi"
+    "\\omega"
+  ]
+
+  greekLetters = greekLetters.concat(greekLetters.map((word) ->  return "\\" + word[1].toUpperCase() + word.substring(2)))
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/snippets/equation_snippets.coffee]: pls create a function `capitalizeWord`

File: lib/ace/ext/papeeria/snippets/equation_snippets.coffee [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+  ])
+
+  greekLetters = greekLetters.map((word) ->
+    caption: word,
+    value: word
+    score: 2
+    meta: "Greek Letter"
+  )
+
+  return SumsAndIntegrals.concat(formulasSnippets.concat(greekLetters))
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/snippets/equation_snippets.coffee]: you really want to sort these entries lexicographically within each group and make it so that groups did not interleave

screenshot 2016-08-15 13 36 46

File: lib/ace/ext/papeeria/tex_completer.coffee [hunk 6]
```diff
@@ -6,16 +6,38 @@ define((require, exports, module) ->
   LatexParsingContext = require("ace/ext/papeeria/latex_parsing_context")
   EQUATION_STATE = PapeeriaLatexHighlightRules.EQUATION_STATE
   LIST_STATE = PapeeriaLatexHighlightRules.LIST_STATE
-  equationEnvironments = [
-    "equation"
-    "equation*"
-  ]
+  equationSnippets = require("./snippets/equation_snippets")
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/tex_completer.coffee]: Please name modules like classes, `EachWordCapitalized`

File: lib/ace/ext/papeeria/tex_completer.coffee [hunk 7]
```diff
@@ -55,66 +78,31 @@ define((require, exports, module) ->
...
-        # it"s temporary fix bug with added \item before \begin{itemize|enumerate}
-        if LatexParsingContext.getContext(editor.session, pos.row) == LIST_STATE && curLine.indexOf("begin") < pos.column
-          editor.insert("\n" + tabString.repeat(indentCount) + "\\item ")
+        cursor = editor.getCursorPosition();
+        line = editor.session.getLine(cursor.row);
+        tabString = editor.session.getTabString();
+        indentString = line.match(/^\s*/)[0];
+        indexOfBegin = line.indexOf("begin")
+
+        if LatexParsingContext.getContext(editor.session, cursor.row) == LIST_STATE &&  indexOfBegin < cursor.column
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/tex_completer.coffee]: any sense to make `getContext` returning a range which starts from \begin start position? This will eliminate the need in this hackety hack with `indexOf "begin"`

File: lib/ace/autocomplete.js [hunk 8]
```diff
@@ -434,7 +434,7 @@ var FilteredList = function(array, filterText) {
         this.filterText = str;
         matches = this.filterCompletions(matches, this.filterText);
         matches = matches.sort(function(a, b) {
-            return b.exactMatch - a.exactMatch || b.score - a.score;
+            return b.exactMatch - a.exactMatch || a.meta > b.meta || b.score - a.score;
```
Comment (dbarashev) [inline lib/ace/autocomplete.js]: pls revert the direction: `b.meta < a.meta` so that it looked consistent with subsequent score difference

File: lib/ace/ext/papeeria/snippets/EquationSnippets.coffee [hunk 9]
```diff
@@ -114,14 +100,14 @@ define((require, exports, module) ->
     "\\varkappa"
   ])
 
-  greekLetters = greekLetters.map((word) ->
+  GREEK_LETTERS = GREEK_LETTERS.map((word) ->
     caption: word,
     value: word
-    score: 2
+    score: 1000
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/snippets/EquationSnippets.coffee]: why don't you assign different score to different letters, in lexicographical order?

File: lib/ace/ext/papeeria/snippets/EquationSnippets.coffee [hunk 10]
```diff
@@ -20,7 +20,7 @@ define((require, exports, module) ->
   SUMS_AND_INTEGRALS = SUMS_AND_INTEGRALS.map((word) ->
       caption: word + "{n}{i=..}{..}",
       snippet: word + "^{${1:n}}_{${2:i=1}}{${3}}"
-      score: 1000
+      score: 1000 - word.charCodeAt(1)
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/snippets/EquationSnippets.coffee]: come on. You have your own list which you can sort and use element indexes.

File: lib/ace/autocomplete.js [hunk 11]
```diff
@@ -434,7 +434,8 @@ var FilteredList = function(array, filterText) {
         this.filterText = str;
         matches = this.filterCompletions(matches, this.filterText);
         matches = matches.sort(function(a, b) {
-            return b.exactMatch - a.exactMatch || a.meta > b.meta || b.score - a.score;
+            return (a.exactMatch < b.exactMatch) || ((b.meta < a.meta) - (a.meta < b.meta))
```
Comment (dbarashev) [inline lib/ace/autocomplete.js]: comparator is supposed to return integer value [[docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort)]

Here you return booleans which may match the expected ints somehow due to crazy JS type conversions, but are extremely fragile and error prone. Please rewrite it so that it explicitly returned integer.

File: lib/ace/ext/papeeria/latex_parsing_context.coffee [hunk 12]
```diff
@@ -9,24 +9,27 @@ define((require, exports, module) ->
...
+        return token.type.split(".").indexOf(type) > -1
+
+    getContext = (session, row, column) ->
+        token = session.getTokenAt(row, column)
+        if token? and isType(token, "math")
+            return EQUATION_STATE
+        else
+            getContextFromRaw(session, row)
+
+    getContextFromRaw = (session, row) ->
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/latex_parsing_context.coffee]: I am not sure if you really meant "raw" (сырой) or "row" (строка), but it you meant the latter then please fix the typo

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 13]
```diff
@@ -147,6 +152,23 @@ define((require, exports, module) ->
...
+            regex : "%.*$"
+        }, {
+            token : "string.math",
+            regex : "\\${1,2}"
+            next  : "start"
+        }, {
+            token : "constant.character.escape.string.math"
+            regex : "\\\\(?:[^a-zA-Z]|[a-zA-Z]+)"
+        }, {
+            token : "error.math",
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: why is it an error?

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 14]
```diff
@@ -147,6 +152,23 @@ define((require, exports, module) ->
         endRule(EQUATION_REGEX)
         endRule(LIST_REGEX)
       ]
+      "math" : [{
+            token : "comment"
+            regex : "%.*$"
+        }, {
+            token : "string.math",
+            regex : "\\${1,2}"
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: any sense to consider also LaTeX-style display math surrounded with `\[` `\]` ?

File: lib/ace/ext/papeeria/snippets/EquationSnippets.coffee [hunk 15]
```diff
@@ -0,0 +1,119 @@
+define((require, exports, module) ->
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/snippets/EquationSnippets.coffee]: CapitalizedNames in CS/JS files are not quite common, at least in Ace and Papeeria code. Pls rename back to `equation_snippets.coffee`

File: lib/ace/ext/papeeria/snippets/EquationSnippets.coffee [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+            """
+      score: 2
+      meta: "Math"
+    }
+    {
+      caption: "\\sqrt{n}"
+      snippet: """
+                \\sqrt{${1:n}}
+            """
+      score: 3
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/snippets/EquationSnippets.coffee]: Please sort Math snippets as well

File: lib/ace/ext/papeeria/snippets/EquationSnippets.coffee (hunk 15 above)
Comment (dbarashev) [inline lib/ace/ext/papeeria/snippets/EquationSnippets.coffee]: CapitalizedNames in CS/JS files are not quite common, at least in Ace and Papeeria code. Pls rename back to `equation_snippets.coffee`

File: lib/ace/ext/papeeria/tex_completer.coffee [hunk 17]
```diff
@@ -143,28 +129,33 @@ define((require, exports, module) ->
       else
         callback(null, @cache)
 
+  compare = (a, b) ->
+    console.log(a.meta, b.meta)
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/tex_completer.coffee]: remove debug output

File: lib/ace/ext/papeeria/tex_completer.coffee [hunk 18]
```diff
@@ -143,28 +129,33 @@ define((require, exports, module) ->
       else
         callback(null, @cache)
 
+  compare = (a, b) ->
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/tex_completer.coffee]: do you use this function anywhere?

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee (hunk 13 above)
Comment (Mikhail-M) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: it's because we should we should get out of this mode.
It is from original latex_highligt_rules.js

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee (hunk 14 above)
Comment (Mikhail-M) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: Why not=)

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee (hunk 13 above)
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: hm, indeed, empty lines are not allowed in display maths. Ok

File: lib/ace/autocomplete.js [hunk 19]
```diff
@@ -434,8 +434,14 @@ var FilteredList = function(array, filterText) {
         this.filterText = str;
         matches = this.filterCompletions(matches, this.filterText);
         matches = matches.sort(function(a, b) {
-            return (a.exactMatch < b.exactMatch) || ((b.meta < a.meta) - (a.meta < b.meta))
-                || ((a.score < b.score) - (b.score < a.score))
+            var result = b.exactMatch - a.exactMatch
```
Comment (dbarashev) [inline lib/ace/autocomplete.js]: pls add semicolons at the end, since this code is JavaScript

File: lib/ace/ext/papeeria/snippets/equation_snippets.coffee [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+      caption: "\\binom{n}{k}"
+      snippet: """
+                \\binom{${1:n}}{${2:k}}
+            """
+      score: 1000
+      meta: "Math"
+    }
+  ]
+
+  MATH_SNIPPETS = MATH_SNIPPETS.sort((a, b) -> a.meta.localeCompare(b.meta))
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/snippets/equation_snippets.coffee]: what? Their meta is always "Math",no?

File: lib/ace/ext/papeeria/snippets/equation_snippets.coffee [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+  MATH_SNIPPETS = for i in [0..MATH_SNIPPETS.length-1]
+      console.log(i)
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/snippets/equation_snippets.coffee]: remove debug logging

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee (hunk 14 above)
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: then please do :)

File: lib/ace/ext/papeeria/snippets/equation_snippets.coffee (hunk 20 above)
Comment (Mikhail-M) [inline lib/ace/ext/papeeria/snippets/equation_snippets.coffee]: oooh
caption should be here

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 22]
```diff
@@ -30,6 +32,14 @@ define((require, exports, module) ->
           stack.push(pushedState)
         return pushedState
 
+    pushStateCheckout = (pushedState, to) ->
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: Any comment on what's going on here? 

Overall, this code needs a dozen of tests.

**Comment by dbarashev** [comment]
This scenario hangs the editor:
add this snippet above \begin{equation} in the demo:

```
\[
x^2
\]
```

When caret is inside `\[\]` it renders a formula which is nice. However, deleting `\]` using backspace if that matters hangs the browser tab

File: lib/ace/ext/papeeria/latex_parsing_context.coffee [hunk 23]
```diff
@@ -9,18 +9,19 @@ define((require, exports, module) ->
...
         return token.type.split(".").indexOf(type) > -1
 
     getContext = (session, row, column) ->
+        state = getContextFromRow(session, row)
         token = session.getTokenAt(row, column)
         if token? and isType(token, "math")
             return EQUATION_STATE
         else
-            getContextFromRow(session, row)
+            return state
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/latex_parsing_context.coffee]: is it true that everyting that changed is the time when `getContextFromRow` is called? How did it help with fixing the crash?

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 24]
```diff
@@ -32,6 +32,7 @@ define((require, exports, module) ->
           stack.push(pushedState)
         return pushedState
 
+    #The same as pushState, but switch our finite state to "to" state
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: with the purpose of what? Comment doesn't need to rephrase hwat is written in the code, we see that the difference is that "to" is returned instead of "pushedState". What's the purpose of this change? When shall Icall pushState and when shall I call pushStateCheckout?

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 25]
```diff
@@ -44,10 +44,16 @@ define((require, exports, module) ->
         return to
 
     popState = (currentState, stack) ->
-      top = stack.pop()
+      if not stack?
+        return 'start'
 
-      if stack?
+      stack.pop()
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: Did you really mean to pop element and forget it? This code will return 'start', right?

```
stack = ['foo']
popState('bar', stack)
```

What's the point?

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 26]
```diff
@@ -44,10 +44,16 @@ define((require, exports, module) ->
...
+      if stack.length == 0
         return 'start'
+
+      top = stack.pop()
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: so if there are more than 1 elements in the stack, you pop twice. Why?

Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: So you silently skip the topmost element and return last but one. WHY?

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 27]
```diff
@@ -152,13 +158,13 @@ define((require, exports, module) ->
 
     # Function for constructing $$ $$ and \[ \] rules
     # regex -- (String) -- regex for definition end of Equation
-    latexMathModeConstructor = (regex) ->
+    latexMathModeConstructor = (regexOfEnd) ->
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: `closingRegex`

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 28]
```diff
@@ -22,16 +27,34 @@ define((require, exports, module) ->
...
     popState = (currentState, stack) ->
-      return stack.pop() or "start"
+      if not stack?
+        return 'start'
+
+      stack.pop()
+      if stack.length == 0
+        return 'start'
+
+      top = stack.pop()
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: `return stack[stack.length - 1]`

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 29]
```diff
@@ -22,16 +27,24 @@ define((require, exports, module) ->
...
-          stack.push(currentState, pushedState)
-        else
-          stack.push(pushedState)
+        stack.push(pushedState)
         return pushedState
 
     popState = (currentState, stack) ->
-      return stack.pop() or "start"
+      if not stack?
+        return 'start'
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: is it expected case? Or something which should never happen? In the latter case throw error

Comment (Mikhail-M) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: it is really incorrect place. 
I'll rewrite it

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 30]
```diff
@@ -36,13 +36,16 @@ define((require, exports, module) ->
 
     popState = (currentState, stack) ->
       if not stack?
-        return 'start'
+        throw new Error("papeeria highlight rules stack error: stack doesn't exist")
+      if not Array.isArray(stack)
+        return "start"
 
-      if currentState? and currentState != stack[stack.length - 1]
+      if not (currentState == "start" and stack.length == 0) && currentState != stack[stack.length - 1]
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: if `stack.length == 0` and `currentState != "start"` then condition to the left of `&&` is evaluated to `true`, you go to the right of `&&` and access stack[-1] which is okay in JS but isnot okay in other languages.

Rewrite it in a simple way:
```
if stack.length == 0
  if currentState != "start" then throw new Error else return "start"

# here we know stack is not empty
if currentState != stack[stack.length-1] then throw new Error
stack.pop()
...
```

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 31]
```diff
@@ -36,12 +36,19 @@ define((require, exports, module) ->
...
+        throw new Error("papeeria_highlight_rules -- stack error: stack doesn't exist")
+
       if not Array.isArray(stack)
         return "start"
 
-      if not (currentState == "start" and stack.length == 0) && currentState != stack[stack.length - 1]
-        throw new Error("papeeria highlight rules stack error, expected " + stack[stack.length - 1] + " found " + currentState)
+      if stack.length == 0
+        if currentState != "start"
+         throw new Error('papeeria_highlight_rules -- stack error: stack should not be empty here')
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: add one more space

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 32]
```diff
@@ -36,12 +36,19 @@ define((require, exports, module) ->
...
+        else return "start"
+
+      # here we know stack is not empty
+      errorMessage = "papeeria_highlight_rules -- stack error: expected " + stack[stack.length - 1] + " found " + currentState
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: is it really true? I thought that `currentState` is the external view of the state and what is actually expected, while the topmost element is what you have found

File: lib/ace/autocomplete.js [hunk 33]
```diff
@@ -434,7 +434,14 @@ var FilteredList = function(array, filterText) {
...
         matches = this.filterCompletions(matches, this.filterText);
         matches = matches.sort(function(a, b) {
-            return b.exactMatch - a.exactMatch || b.score - a.score;
+            var result = a.exactMatch - b.exactMatch
+            if (result != 0)
+                return result
+            result = a.meta.localeCompare(b.meta)
+            if (result != 0)
+                return result
+            result = b.score - a.score
```
Comment (dbarashev) [inline lib/ace/autocomplete.js]: This looks like a bug (reverted order comparing to the previous comparisons) unless you know that the score value is reverted as well. Please swap b and a here and set scores to `i` instead of `1000 - i` where you sort the completions

File: lib/ace/ext/papeeria/latex_parsing_context.coffee [hunk 34]
```diff
@@ -4,29 +4,38 @@ define((require, exports, module) ->
     PapeeriaLatexHighlightRules = require("ace/ext/papeeria/papeeria_latex_highlight_rules")
     EQUATION_STATE = PapeeriaLatexHighlightRules.EQUATION_STATE
     LIST_STATE = PapeeriaLatexHighlightRules.LIST_STATE
+
+    EQUATION_TOKENTYPE  = PapeeriaLatexHighlightRules.EQUATION_TOKENTYPE
+    LIST_TOKENTYPE  = PapeeriaLatexHighlightRules.LIST_TOKENTYPE
+
     ###
      * @param {Number} row
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/latex_parsing_context.coffee]: This comment is not valid anymore

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 35]
```diff
@@ -8,8 +8,13 @@ define((require, exports, module) ->
   EQUATION_REGEX = "equation|equation\\*"
   LIST_STATE = "list"
   EQUATION_STATE = "equation"
+  LIST_TOKENTYPE = 'list'
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: nit: use double quotes?

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 36]
```diff
@@ -22,16 +27,35 @@ define((require, exports, module) ->
...
+        stack.push(pushedState)
         return pushedState
 
     popState = (currentState, stack) ->
-      return stack.pop() or "start"
+      if not stack?
+        throw new Error("papeeria_highlight_rules -- stack error: stack doesn't exist")
+
+      if not Array.isArray(stack)
+        return "start"
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: check that stack is actually a string with value "start" ?

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 37]
```diff
@@ -22,16 +27,35 @@ define((require, exports, module) ->
...
+          throw new Error('papeeria_highlight_rules -- stack error: stack should not be empty here')
+        else
+          return "start"
+
+      # here we know stack is not empty
+      errorMessage = "papeeria_highlight_rules -- stack error: expected " + currentState + " found " + stack[stack.length - 1]
+      if currentState != stack[stack.length-1] then throw new Error(errorMessage)
+
+      stack.pop()
+      if stack.length == 0
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: This will never be true because there is a check at line 44

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 38]
```diff
@@ -121,6 +156,32 @@ define((require, exports, module) ->
         next: popState
       }
 
+
+    # Function for constructing $$ $$ and \[ \] rules
+    # regex -- (String) -- regex for definition end of Equation
+    latexMathModeConstructor = (regexOfEnd) ->
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: s/regexOfEnd/closingRegex

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 39]
```diff
@@ -139,15 +200,54 @@ define((require, exports, module) ->
         endRule(EQUATION_REGEX)
         endRule(LIST_REGEX)
 
+        # will be simplified
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: will it?

File: lib/ace/ext/papeeria/snippets/equation_snippets.coffee [hunk 40]
```diff
@@ -0,0 +1,125 @@
...
+      meta: "Sums and integrals"
+
+
+  MATH_SNIPPETS = [
+    {
+      caption: "\\frac{num}{denom}"
+      snippet: """
+                \\frac{${1:num}}{${2:denom}}
+            """
+      score: 1000
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/snippets/equation_snippets.coffee]: you don't this score anymore, do you?

File: lib/ace/ext/papeeria/snippets/equation_snippets.coffee [hunk 41]
```diff
@@ -0,0 +1,125 @@
...
+  GREEK_LETTERS = for i in [0..GREEK_LETTERS.length]
+    caption: GREEK_LETTERS[i]
+    value: GREEK_LETTERS[i]
+    score: 1000 - i
+    meta: "Greek Letter"
+
+
+  return SUMS_AND_INTEGRALS.concat(MATH_SNIPPETS, GREEK_LETTERS)
+);
+
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/snippets/equation_snippets.coffee]: nit: remove blank lines

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee (hunk 36 above)
Comment (Mikhail-M) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: okey

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee (hunk 37 above)
Comment (Mikhail-M) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: why?
We have stack.pop in previous line

Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: right, missed that line

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 42]
```diff
@@ -39,6 +39,9 @@ define((require, exports, module) ->
         throw new Error("papeeria_highlight_rules -- stack error: stack doesn't exist")
 
       if not Array.isArray(stack)
+        if stack != 'start'
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: double quotes

File: lib/ace/autocomplete.js [hunk 43]
```diff
@@ -434,7 +434,14 @@ var FilteredList = function(array, filterText) {
         this.filterText = str;
         matches = this.filterCompletions(matches, this.filterText);
         matches = matches.sort(function(a, b) {
-            return b.exactMatch - a.exactMatch || b.score - a.score;
+            var result = a.exactMatch - b.exactMatch
```
Comment (dbarashev) [inline lib/ace/autocomplete.js]: revert this too?

File: lib/ace/ext/papeeria/latex_parsing_context.coffee [hunk 44]
```diff
@@ -4,29 +4,38 @@ define((require, exports, module) ->
...
      *
      * Returns context at row.
     ###
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/latex_parsing_context.coffee]: comment is not valid anymore again

File: lib/ace/autocomplete.js [hunk 45]
```diff
@@ -434,7 +434,19 @@ var FilteredList = function(array, filterText) {
         this.filterText = str;
         matches = this.filterCompletions(matches, this.filterText);
         matches = matches.sort(function(a, b) {
-            return b.exactMatch - a.exactMatch || b.score - a.score;
+            var result = b.exactMatch - a.exactMatch
+            if (result != 0)
+                return result
+            if (a.meta_score !== null && typeof a.meta_score !== 'undefined' &&
+                b.meta_score !== null && typeof b.meta_score !== 'undefined')
+                result = a.meta_score - b.meta_score
```
Comment (dbarashev) [inline lib/ace/autocomplete.js]: lets keep the meaning of score consistent. Bigger score => higher in the list

Comment (Mikhail-M) [inline lib/ace/autocomplete.js]: lets

File: lib/ace/ext/papeeria/tex_completer.coffee [hunk 46]
```diff
@@ -4,48 +4,93 @@ define((require, exports, module) ->
...
 
-  basicSnippets = [
+  EQUATION_ENVIRONMENTS = [
+      "equation"
+      "equation*"
+  ]
+
+
+  EQUATION_ENV_SNIPPETS = for env in EQUATION_ENVIRONMENTS
+    {
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/tex_completer.coffee]: remove curly brackets? Or add them for GREEK_LETTRS and MATH_SNIPPETS?

**Comment by Mikhail-M** [comment]
it is all?

**Comment by dbarashev** [comment]
Cool. Final remarks: sort base snippets too, and give them lower meta score. In the body of the document it is more natural to use \begin{enumeration} rather than \usepackage

File: lib/ace/ext/papeeria/tex_completer.coffee [hunk 47]
```diff
@@ -162,8 +170,9 @@ define((require, exports, module) ->
           @refGetter.getReferences(@referencesUrl, callback)
         else switch context
           when "start" then callback(null, BASIC_SNIPPETS.concat(LIST_SNIPPET,
-            EQUATION_ENV_SNIPPETS, REFERENCE_SNIPPET))
-          when LIST_STATE then callback(null, LIST_KEYWORDS.concat(EQUATION_ENV_SNIPPETS, LIST_SNIPPET, LIST_END_ENVIRONMENT))
+            EQUATION_ENV_SNIPPETS, REFERENCE_SNIPPET, CITATION_SNIPPET))
+          when LIST_STATE then callback(null, LIST_KEYWORDS.concat(EQUATION_ENV_SNIPPETS,
+            LIST_SNIPPET, LIST_END_ENVIRONMENT, CITATION_SNIPPET, REFERENCE_SNIPPET))
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/tex_completer.coffee]: please use the same order of lists in both concatenations (unless the order must be exactly like it is now).

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 48]
```diff
@@ -187,75 +213,42 @@ define((require, exports, module) ->
 
     # For unknown reasons  we can"t use constants in block below, because background_tokenizer
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: is tis comment still actual? Any sense to use constants as keys below (and probably booleans as values because current values seem to be the same as keys)

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee [hunk 49]
```diff
@@ -60,11 +60,26 @@ define((require, exports, module) ->
 
       return stack[stack.length - 1]
 
-    basicRules = [
+    #specialized rules for context
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: add a space after `#`

File: lib/ace/ext/papeeria/tex_completer.coffee (hunk 47 above)
Comment (Mikhail-M) [inline lib/ace/ext/papeeria/tex_completer.coffee]: the order does not depend on this concatenation, couse we sort this by autocomplete.js

File: lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee (hunk 48 above)
Comment (Mikhail-M) [inline lib/ace/ext/papeeria/papeeria_latex_highlight_rules.coffee]: I think this comment was not actual never

File: lib/ace/ext/papeeria/tex_completer.coffee (hunk 47 above)
Comment (dbarashev) [inline lib/ace/ext/papeeria/tex_completer.coffee]: Exactly. Computer doesn't care. It is human who does.

Comment (dbarashev) [inline lib/ace/ext/papeeria/tex_completer.coffee]: @Mikhail-M to be clear, please do reorder concat arguments so that the difference between them was obvious to human. The same order of cmmon arguments in both concat calls.

**Comment by dbarashev** [comment]
LGTM
